File: packages/python/src/ghagen/pin/update.py
"""Apply version updates to user Python source files.

Replaces ``uses=`` string literals in-place, scoped to files identified
by the source-tracking module.
"""

from __future__ import annotations

from pathlib import Path


def apply_updates(
    updates: dict[str, str],
    ref_locations: dict[str, list[Path]],
) -> list[Path]:
    """Replace ``uses`` strings in user source files.

    Args:
        updates: Mapping of ``old_uses → new_uses``
            (e.g. ``{"actions/checkout@v4": "actions/checkout@v5"}``).
        ref_locations: Mapping of ``uses → [files]`` from
            :func:`ghagen.pin.sources.locate_uses_refs`.  Only files
            listed here are modified.

    Returns:
        Sorted list of files that were actually changed.
    """
    changed: set[Path] = set()

    for old_uses, new_uses in updates.items():
        if old_uses == new_uses:
            continue
        files = ref_locations.get(old_uses)
        if not files:
            continue
        for path in files:
            content = path.read_text()
            if old_uses in content:
                path.write_text(content.replace(old_uses, new_uses))
                changed.add(path)

    return sorted(changed)

File: packages/typescript/src/pin/update.ts
/**
 * Apply version updates to user TS/JS source files.
 *
 * Replaces `uses` string literals in-place, scoped to files identified
 * by the source-tracking module.
 */

import { readFileSync, writeFileSync } from "node:fs";

/**
 * Replace `uses` strings in user source files.
 *
 * Returns the sorted list of files that were actually changed.
 */
export function applyUpdates(
  updates: ReadonlyMap<string, string> | Readonly<Record<string, string>>,
  refLocations: ReadonlyMap<string, readonly string[]> | Readonly<Record<string, readonly string[]>>,
): string[] {
  const updatesMap =
    updates instanceof Map ? updates : new Map(Object.entries(updates));
  const locMap =
    refLocations instanceof Map
      ? refLocations
      : new Map(Object.entries(refLocations));

  const changed = new Set<string>();
  for (const [oldUses, newUses] of updatesMap.entries()) {
    if (oldUses === newUses) continue;
    const files = locMap.get(oldUses);
    if (!files || files.length === 0) continue;
    for (const path of files) {
      const content = readFileSync(path, "utf8");
      if (content.includes(oldUses)) {
        writeFileSync(path, content.split(oldUses).join(newUses));
        changed.add(path);
      }
    }
  }

  return [...changed].sort();
}
